cs/src/core/Index/FASTER/FASTER.cs

```diff
@@ -319,6 +319,108 @@ internal Status ContextRead(ref Key key, ref Input input, ref Output output, Con
+            ClientSession<Key, Value, Input, Output, Context, Functions> clientSession,
+            ref Key key, ref Input input, Context context = default, CancellationToken token = default)
+        {
+            var pcontext = default(PendingContext);
+            Output output = default;
+
+            var nextSerialNum = clientSession.ctx.serialNum + 1;
+
+            OperationStatus internalStatus;
+
+            if (clientSession.SupportAsync) clientSession.UnsafeResumeThread();
+            try
+            {
+                do
+                {
+                    internalStatus = InternalRead(ref key, ref input, ref output,
+                    ref context, ref pcontext, clientSession.ctx, nextSerialNum);
+                }
+                while (internalStatus == OperationStatus.CPR_SHIFT_DETECTED);
+            }
+            finally
+            {
+                if (clientSession.SupportAsync) clientSession.UnsafeSuspendThread();
+            }
+
+            switch (internalStatus)
+            {
+                case OperationStatus.SUCCESS:
+                case OperationStatus.NOTFOUND:
+
+                    clientSession.ctx.serialNum = nextSerialNum;
+
+                    return new ValueTask<(Status, Output)>(((Status)internalStatus, output));
+
+                case OperationStatus.RECORD_ON_DISK:
+                    return SlowReadAsync(this, clientSession, pcontext, nextSerialNum, token);
+                    
+                
+                default:
+                    throw new Exception($"Unexpected {nameof(OperationStatus)} while reading => {internalStatus}");
+
+            }
+
+            static async ValueTask<(Status, Output)> SlowReadAsync(
+                FasterKV<Key, Value, Input, Output, Context, Functions> @this,
+                ClientSession<Key, Value, Input, Output, Context, Functions> clientSession,
+                PendingContext pendingContext, long nextSerialNum, CancellationToken token = default)
+            {
+
+                var diskRequest = @this.ScheduleGetFromDisk(clientSession.ctx, ref pendingContext);
+
+                clientSession.ctx.serialNum = nextSerialNum;
+
+                try
+                {
+
+                    clientSession.ctx.pendingReads.Add(pendingContext.id, pendingContext);
+
+                    token.ThrowIfCancellationRequested();
+
+                    if (@this.epoch.ThisInstanceProtected())
+                        throw new NotSupportedException("Async operations not supported over protected epoch");
+
+                    diskRequest = await diskRequest.asyncOperation.ValueTaskOfT;
+
+                    lock (clientSession.ctx)
+                    {
+
+                        if (clientSession.SupportAsync) clientSession.UnsafeResumeThread();
+                        try
+                        {
+                            #region RelaxedCPR
+                            if (!@this.RelaxedCPR)
```

Comment:
For async code ONLY: we can assume and assert that RelaxedCPR is true. This should help simplify the code (including getting git of release of latch in InternalCompleteIOPendingReadRequestAsync).

Reply:
Code replaced with a ```Debug.Assert(@this.RelaxedCPR);```